Add tutorial about Single Page Application (#1514)
yewstack/yew: website/versioned_docs/version-0.17.3/getting-started/tutorials/single-page-application.md

---
title: Single page application
sidebar_label: Single page application
description: Learn to write a SPA
id: version-0.17.3-single-page-application
original_id: single-page-application
---

> This tutorial is based on an tutorial by [Sheshbabu](https://github.com/sheshbabu). The original article can be found [here.](http://www.sheshbabu.com/posts/rust-wasm-yew-single-page-application/). It has been slightly reworked and should give a pretty good understanding of Yew and how to build single pages Application. Still, if you see a way in which it can improve this tutorial to do reach the Yew community. Contributions are welcome.

Author: It just a way to encourage people to do a PR to improve the tutorial if they see a way to do so. How you would phrase it?

                </div>
            }
        }
    }
}
```

Very similar to the HomePage component. Let's also add this file to the module tree:

```diff
  // src/pages/mod.rs
  mod home;
+ mod product_detail;

  pub use home::Home;
+ pub use product_detail::ProductDetail;
```

This is how it looks like:

![](/img/tutorials/single-page-application/image-5.png)

We can now move between multiple pages without refreshing the page!

## State Management

One thing you might have noticed in the `ProductDetail` page is that clicking on the "Add To Cart" button doesn't update the cart. This is because the state that holds the list of products in cart `cart_products` currently resides inside `Home` page component:
![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-6.png)

To share state between two components, we can either:

- Hoist the state to a common ancestor
- Move state to global app state

The `App` component is a common ancestor to both `ProductDetail` and `Home`. We can move the `cart_products` state there and pass it as props to `ProductDetail` and `Home`.

![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-7.png)

> This works fine for shallow component hierarchies but when you have a deep component hierarchy (which is common in larger SPAs), you will need to pass this state through multiple layers of components (which might not have use for this prop) to reach the desired node. This is called "Prop Drilling".

Reviewer: ```suggestion
This works fine for shallow component hierarchies but when you have a deep component hierarchy (which is common in larger SPAs), you will need to pass this state through multiple layers of components (which might not have use for this prop) to reach the desired node. This is called "prop drilling".

```

You can see that `cart_products` is now passed from `App` to `AddToCart` component via `ProductDetail` and `Home` even though they have no use for this state. Imagine the same scenario with components many layers deep.

This is the problem the global state solves. here is how it would look like:

![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-8.png)

Notice how there is a direct link between the components that need this state and the global state.

> Unfortunately, Yew doesn't seem to have a [good solution](https://github.com/yewstack/yew/issues/576) for this. Today, the recommended solution is to use `Agents` for broadcasting state changes via pubsub. Yet, this approach might be hard to reason around. The story around state management in Yew is far from definitive and solutions to this problem may appear in the future.

Reviewer: ```suggestion
Unfortunately, Yew doesn't seem to have a [good solution](https://github.com/yewstack/yew/issues/576) for this. Today, the recommended solution is to use `Agents` for broadcasting state changes via pubsub. Yet, this approach might be hard to reason about. The story around state management in Yew is far from definitive and solutions to this problem may appear in the future.

```

![](/img/tutorials/single-page-application/rust-wasm-yew-single-page-application-9.png)

```rust
// src/components/navbar.rs
use crate::types::CartProduct;
use yew::prelude::*;

pub struct Navbar {
    props: Props,
}

#[derive(Properties, Clone)]
pub struct Props {
    pub cart_products: Vec<CartProduct>,
}

impl Component for Navbar {
    type Message = ();
    type Properties = Props;

    fn create(props: Self::Properties, _link: ComponentLink<Self>) -> Self {
        Self { props }
    }

    fn update(&mut self, _msg: Self::Message) -> ShouldRender {
        true
    }

    fn change(&mut self, props: Self::Properties) -> ShouldRender {
        self.props = props;
        true
    }

    fn view(&self) -> Html {
        let cart_value = self
            .props
            .cart_products
            .iter()
            .fold(0.0, |acc, cp| acc + (cp.quantity as f64 * cp.product.price));

        html! {
            <div class="navbar">
                <div class="navbar_title">{"RustMart"}</div>
              <div class="navbar_cart_value">{format!("${:.2}", cart_value)}</div>
            </div>
        }
    }
}
```

> Notice how we started using the `change` lifecycle methods in the `Navbar` component. When the props sent from parent changes, we need to update the props inside the component so the UI re-renders.

Reviewer: ```suggestion
Notice how we started using the `change` lifecycle methods in the `Navbar` component. When the props sent from the parent change, we need to update the props inside the component so that the UI re-renders.

```

              .iter()
              .map(|product: &Product| {
-                 let product_id = product.id;
                  html! {
-                   <ProductCard product={product} on_add_to_cart=self.link.callback(move |_| Msg::AddToCart(product_id))/>
+                   <ProductCard product={product} on_add_to_cart=self.props.on_add_to_cart.clone()/>
                  }
              })
              .collect();

-        let cart_value = self
-            .state
-            .cart_products
-            .iter()
-            .fold(0.0, |acc, cp| acc + (cp.quantity as f64 * cp.product.price));

          if !self.state.get_products_loaded {
              // No changes
          } else if let Some(_) = self.state.get_products_error {
              // No changes
          } else {
              html! {
-               <div>
-                 <div class="navbar">
-                     <div class="navbar_title">{"RustMart"}</div>
-                     <div class="navbar_cart_value">{format!("${:.2}", cart_value)}</div>
-                 </div>
                  <div class="product_card_list">{products}</div>
-               </div>
              }
          }
      }
  }
```

> Now we can finally add to cart from `ProductDetail` page and we can also see the navbar in all pages

Reviewer: Not 100% sure what the semantics of this sentence should look like but something like this (depending on the intended meaning of the sentence) might do the trick.
```suggestion
Now we can finally add a link to the "add to cart" component from the `ProductDetail` page and we can also see the navbar on all pages

```
